src/registry.rs

```diff
@@ -0,0 +1,243 @@
+    /// Create a new [`DefaultObjectStoreRegistry`] with no registered stores
+    pub fn new() -> Self {
+        let object_stores = RwLock::new(HashMap::new());
+        Self { object_stores }
+    }
+}
+
+///
+/// Stores are registered based on the scheme, host and port of the provided URL.
+///
+/// For example:
+///
+/// - `file:///my_path` will return the default LocalFS store
+/// - `s3://bucket/path` will return a store registered with `s3://bucket` if any
+/// - `hdfs://host:port/path` will return a store registered with `hdfs://host:port` if any
+impl ObjectStoreRegistry for DefaultObjectStoreRegistry {
+    fn register_store(
+        &self,
+        url: &Url,
+        store: Arc<dyn ObjectStore>,
+    ) -> Option<Arc<dyn ObjectStore>> {
+        let s = get_url_key(url);
+        let mut stores = self.object_stores.write().unwrap();
+        stores.insert(s, store)
+    }
+
+    fn get_store(&self, url: &Url) -> Option<Arc<dyn ObjectStore>> {
+        let s = get_url_key(url);
+        let mut stores = self.object_stores.write().unwrap();
+
+        if let Some(store) = stores.get(&s) {
+            return Some(Arc::clone(store));
+        }
+
+        match parse_url(url) {
+            Ok((store, _)) => {
+                let store = Arc::from(store);
+                stores.insert(s, Arc::clone(&store));
+                Some(store)
+            }
+            Err(_) => None,
+        }
+    }
+
+    fn list_urls(&self) -> Vec<String> {
+        let stores = self.object_stores.read().unwrap();
+        stores.keys().cloned().collect()
+    }
+}
+
+/// Get the key of a url for object store registration.
+/// The credential info will be removed
+fn get_url_key(url: &Url) -> String {
+    format!(
+        "{}://{}",
+        url.scheme(),
+        &url[url::Position::BeforeHost..url::Position::AfterPort],
+    )
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_get_url_key() {
+        let file = Url::parse("file://").unwrap();
+        let key = get_url_key(&file);
+        assert_eq!(key.as_str(), "file://");
+
+        let url = Url::parse("s3://bucket").unwrap();
+        let key = get_url_key(&url);
+        assert_eq!(key.as_str(), "s3://bucket");
+
+        let url = Url::parse("s3://username:password@host:123").unwrap();
+        let key = get_url_key(&url);
+        assert_eq!(key.as_str(), "s3://host:123");
+    }
+
+    #[cfg(not(target_arch = "wasm32"))]
+    #[test]
+    fn test_register_store() {
+        use crate::local::LocalFileSystem;
+        let registry = DefaultObjectStoreRegistry::new();
+        let url = Url::parse("file:///foo/bar").unwrap();
+        let store = Arc::new(LocalFileSystem::new()) as Arc<dyn ObjectStore>;
+        let old_store = registry.register_store(&url, Arc::clone(&store));
+        assert!(
+            old_store.is_none(),
+            "Should not return a previous store when registering a new one"
+        );
+        let retrieved_store = registry
```

Comment:
I think it would be good to test that a sub-prefix of this worked too -- aka test that `file:///foo/bar/baz` also resolves to the store

Also it would be good to test that `file://bar` did not resolve

Reply:
Added here: 005c9f2